src/controllers/router/index.tsx

```diff
@@ -1,77 +1,89 @@
-import { getRouterState, RouterContainer } from '../router-store';
-import { UnlistenHistory } from '../router-store/types';
+import { getResourceStore, ResourceContainer } from '../resource-store';
+import {
+  getRouterState,
+  RouterContainer,
+  getRouterStore,
+} from '../router-store';
 
-import { RouterProps } from './types';
+import {
+  RouterProps,
+  MemoryRouterProps,
+  RequestResourcesParams,
+} from './types';
 
-/**
- * Default prop provider for the RouterContainer.
- *
- */
-export class Router extends Component<RouterProps> {
-  static defaultProps = {
-    isStatic: false,
-    isGlobal: true,
-    history: DEFAULT_HISTORY,
-  };
-
-  /**
-   * Keep a copy of the history listener so that we can be sure that
-   * on unmount we call the listener that was in the router store at the
-   * time of mounting.
-   *
-   * This prevents an issue where the wrong listener is removed if the router
-   * is re-mounted.
-   */
-  unlistenHistory: UnlistenHistory | null = null;
-
-  componentDidMount() {
-    if (!this.props.isStatic) {
-      const state = getRouterState();
-      this.unlistenHistory = state.unlisten;
-    }
-  }
-
-  /**
-   * Ensures that the router store stops listening to history when the Router
-   * is unmounted.
-   */
-  componentWillUnmount() {
-    if (this.unlistenHistory) {
-      this.unlistenHistory();
-    }
-  }
+export const Router = ({
+  basePath,
+  children,
+  history,
+  initialRoute,
+  isGlobal = true,
+  onPrefetch,
+  resourceContext,
+  resourceData,
+  routes,
+}: RouterProps) => {
+  useEffect(() => {
+    const { unlisten } = getRouterState();
 
-  render() {
-    const {
-      children,
-      routes,
-      history,
-      initialRoute,
-      isStatic,
-      isGlobal,
-      basePath,
-      resourceContext,
-      resourceData,
-      onPrefetch,
-    } = this.props;
+    return () => {
+      unlisten && unlisten();
+    };
+  }, []);
 
-    return (
+  return (
+    <ResourceContainer isGlobal={isGlobal}>
       <RouterContainer
         basePath={basePath}
-        routes={routes}
         history={history}
         initialRoute={initialRoute}
-        isStatic={isStatic}
+        isGlobal={isGlobal}
+        onPrefetch={onPrefetch}
         resourceContext={resourceContext}
         resourceData={resourceData}
-        onPrefetch={onPrefetch}
-        isGlobal={isGlobal}
+        routes={routes}
       >
         {children}
       </RouterContainer>
-    );
-  }
+    </ResourceContainer>
+  );
+};
+
+/**
+ * The entry point for requesting resource data on the server.
+ * Pass the result data into the router as a prop in order to hydrate it.
+ */
+Router.requestResources = async ({
+  location,
+  history,
+  timeout,
+  ...bootstrapProps
+}: RequestResourcesParams) => {
+  const { bootstrapStore, requestRouteResources } = getRouterStore().actions;
+
+  bootstrapStore({
+    ...bootstrapProps,
+    history: history || createMemoryHistory({ initialEntries: [location] }),
```

Comment:
If `history` is now required as prop, should we mandate it here too? Wonder if this might cause problem if we pass a history v4 and consumer is using history v5

Reply:
It's a peer dependency, so it shouldn't matter. But I can change it if you like

Reply:
We said we wanna deprecate this API later on anyway, so fine to leave as is